File: source/ts/src/astronomy/moonPhase.ts
import { searchMoonQuarter, nextMoonQuarter } from './lunation';
import { validateDate, validateLocalYearWindow } from '../utils/validation';
import { localYearWindow, clampToSupported } from '../utils/timezone';

/** The four principal lunar phases: precise instants, not the ~24h tithi windows. */
export type MoonPhaseName = 'new' | 'first_quarter' | 'full' | 'last_quarter';

export interface MoonPhaseEvent {
  phase: MoonPhaseName;
  /** UTC instant; only the calendar date it lands on is local. */
  time: Date;
}

const QUARTER_TO_PHASE: readonly MoonPhaseName[] = [
  'new', 'first_quarter', 'full', 'last_quarter',
];

/** Every lunar-phase event whose instant falls within `[start, end]` inclusive, ascending. */
export function computeMoonPhasesInRange(start: Date, end: Date): MoonPhaseEvent[] {
  validateDate(start);
  validateDate(end);
  if (start.getTime() > end.getTime()) {
    throw new RangeError(
      `start (${start.toISOString()}) must be ≤ end (${end.toISOString()})`,
    );
  }

  const endMs = end.getTime();
  const out: MoonPhaseEvent[] = [];

  let mq = searchMoonQuarter(new Date(start.getTime() - 31 * 24 * 3600_000));
  const maxSteps = Math.ceil((endMs - start.getTime()) / (24 * 3600_000) / 6) + 60;
  for (let step = 0; step < maxSteps; step++) {
    const t = mq.time;
    const ms = t.getTime();
    if (ms > endMs) break;
    if (ms >= start.getTime()) {
      out.push({ phase: QUARTER_TO_PHASE[mq.quarter]!, time: t });
    }
    mq = nextMoonQuarter(mq);
  }
  return out;
}

/** Every principal phase whose instant falls in local calendar year `year`, each boundary at its own offset. */
export function computeMoonPhasesForYear(
  year: number,
  options: { timezone: number | string },
): MoonPhaseEvent[] {
  const [start, end] = localYearWindow(year, options.timezone);
  validateLocalYearWindow(year, start, end);
  return computeMoonPhasesInRange(new Date(clampToSupported(start)), new Date(clampToSupported(end)));
}

/** @deprecated Renamed to {@link computeMoonPhasesInRange} in v5. */
export const getMoonPhasesInRange = computeMoonPhasesInRange;

File: source/go/internal/astronomy/moonphase.go
package astronomy

import (
	"context"
	"math"

	"github.com/ishankgupta95/panchang/source/go/v5/internal/utils"
	"github.com/ishankgupta95/panchang/source/go/v5/types"
)

var quarterToPhase = [4]MoonPhaseName{
	MoonPhaseNew, MoonPhaseFirstQuarter, MoonPhaseFull, MoonPhaseLastQuarter,
}

func ComputeMoonPhasesInRange(ctx context.Context, eph *EphemerisCtx, startMs, endMs int64) ([]MoonPhaseEvent, error) {
	if err := ctx.Err(); err != nil {
		return nil, err
	}
	if err := utils.ValidateDate(startMs); err != nil {
		return nil, err
	}
	if err := utils.ValidateDate(endMs); err != nil {
		return nil, err
	}
	if startMs > endMs {
		return nil, types.Codef(types.ErrInvalidInput,
			"start (%s) must be <= end (%s)", isoString(startMs), isoString(endMs))
	}

	out := []MoonPhaseEvent{}

	mq, err := SearchMoonQuarter(eph, startMs-31*24*3600_000)
	if err != nil {
		return nil, err
	}
	maxSteps := int(math.Ceil(float64(endMs-startMs)/(24*3600_000)/6)) + 60
	for step := 0; step < maxSteps; step++ {
		if err := ctx.Err(); err != nil {
			return nil, err
		}
		ms := mq.TimeMs
		if ms > endMs {
			break
		}
		if ms >= startMs {
			out = append(out, MoonPhaseEvent{Phase: quarterToPhase[mq.Quarter], TimeMs: types.Date(ms)})
		}
		mq, err = NextMoonQuarter(eph, mq)
		if err != nil {
			return nil, err
		}
	}
	return out, nil
}

func ComputeMoonPhasesForYear(ctx context.Context, eph *EphemerisCtx, year int, options MoonPhasesForYearOptions) ([]MoonPhaseEvent, error) {
	if err := ctx.Err(); err != nil {
		return nil, err
	}
	startUtc, endUtc, err := utils.LocalYearWindow(year, options.Timezone)
	if err != nil {
		return nil, err
	}
	if err := utils.ValidateLocalYearWindow(year, startUtc, endUtc); err != nil {
		return nil, err
	}
	return ComputeMoonPhasesInRange(ctx, eph, utils.ClampToSupported(startUtc), utils.ClampToSupported(endUtc))
}

func GetMoonPhasesInRange(ctx context.Context, eph *EphemerisCtx, startMs, endMs int64) ([]MoonPhaseEvent, error) {
	return ComputeMoonPhasesInRange(ctx, eph, startMs, endMs)
}
